dds/DCPS/RTPS/Spdp.cpp

```diff
@@ -983,6 +1003,123 @@ Spdp::send_handshake_request(const DCPS::RepoId& guid, DiscoveredParticipant& dp
+  }
+
+  if (!from_discovery && dp.handshake_state_ != DCPS::HANDSHAKE_STATE_DONE) {
+    // Ignore auth reqs when already in progress.
+    return;
+  }
+
+  // Reset.
+  purge_handshake_deadlines(iter);
+  dp.handshake_deadline_ = DCPS::MonotonicTimePoint::now() + config_->max_auth_time();
+  handshake_deadlines_.insert(std::make_pair(dp.handshake_deadline_, guid));
+  tport_->handshake_deadline_processor_->schedule(config_->max_auth_time());
+
+  DDS::Security::Authentication_var auth = security_config_->get_authentication();
+  DDS::Security::SecurityException se = {"", 0, 0};
+
+  const DDS::Security::ValidationResult_t vr = auth->validate_remote_identity(dp.identity_handle_,
+                                                                              dp.local_auth_request_token_, dp.remote_auth_request_token_, identity_handle_, dp.identity_token_, guid, se);
+
+  dp.have_auth_req_msg_ = !(dp.local_auth_request_token_ == DDS::Security::Token());
+  if (dp.have_auth_req_msg_) {
+    dp.auth_req_msg_.message_identity.source_guid = guid_;
+    dp.auth_req_msg_.message_identity.sequence_number = static_cast<CORBA::LongLong>((++stateless_sequence_number_).getValue());
+    dp.auth_req_msg_.message_class_id = DDS::Security::GMCLASSID_SECURITY_AUTH_REQUEST;
+    dp.auth_req_msg_.destination_participant_guid = guid;
+    dp.auth_req_msg_.destination_endpoint_guid = GUID_UNKNOWN;
+    dp.auth_req_msg_.source_endpoint_guid = GUID_UNKNOWN;
+    dp.auth_req_msg_.related_message_identity.source_guid = GUID_UNKNOWN;
+    dp.auth_req_msg_.related_message_identity.sequence_number = 0;
+    dp.auth_req_msg_.message_data.length(1);
+    dp.auth_req_msg_.message_data[0] = dp.local_auth_request_token_;
+    // Send the auth req immediately to reset the remote if they are
+    // still authenticated with us.
+    if (sedp_.write_stateless_message(dp.auth_req_msg_, make_id(guid, ENTITYID_P2P_BUILTIN_PARTICIPANT_STATELESS_READER)) != DDS::RETCODE_OK) {
+      if (DCPS::security_debug.auth_debug) {
+        ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) {auth_debug} Spdp::attempt_authentication() - ")
+                   ACE_TEXT("Unable to write auth req message retry.\n")));
+      }
+    } else {
+      if (DCPS::security_debug.auth_debug) {
+        ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) {auth_debug} DEBUG: Spdp::attempt_authentication() - ")
+                   ACE_TEXT("Sent auth req message for participant: %C\n"),
+                   OPENDDS_STRING(DCPS::GuidConverter(guid)).c_str()));
+      }
+    }
+    // Schedule resend.
+    schedule_handshake_resend(config_->auth_resend_period(), guid);
+  }
+
+  switch (vr) {
+  case DDS::Security::VALIDATION_OK: {
+    dp.auth_state_ = DCPS::AUTH_STATE_AUTHENTICATED;
+    dp.handshake_state_ = DCPS::HANDSHAKE_STATE_DONE;
+    purge_handshake_deadlines(participants_.find(guid));
```

Comment:
```suggestion
    purge_handshake_deadlines(iter);
```
I think there are a few of these